File: apps/server/agent/src/compaction/cut-point.ts
import type { AgentMessage } from "@/agent/src/types/index.js";
import { estimateMessageTokens } from "./token-estimator.js";

export interface CutPointResult {
  /** Index of the first message to keep in the active context. */
  firstKeptIndex: number;
  /** Whether the cut point successfully landed at a clean user boundary. */
  isUserBoundary: boolean;
}

/**
 * Determine whether slicing at `candidateIndex` would orphan any toolResult
 * from its corresponding assistant toolCall.
 */
export function isToolCallSafe(messages: AgentMessage[], candidateIndex: number): boolean {
  if (candidateIndex <= 0 || candidateIndex >= messages.length) {
    return true;
  }

  // Never cut at a toolResult
  if (messages[candidateIndex].role === "toolResult") {
    return false;
  }

  // Collect all tool call IDs produced in the discarded prefix (0 .. candidateIndex - 1)
  const discardedToolCallIds = new Set<string>();
  for (let i = 0; i < candidateIndex; i++) {
    const msg = messages[i];
    if (msg.role === "assistant") {
      for (const part of msg.content) {
        if (part.type === "toolCall") {
          discardedToolCallIds.add(part.call.id);
        }
      }
    }
  }

  // Ensure no kept message in (candidateIndex .. end) is a toolResult for a discarded call
  for (let i = candidateIndex; i < messages.length; i++) {
    const msg = messages[i];
    if (msg.role === "toolResult") {
      for (const res of msg.results) {
        if (discardedToolCallIds.has(res.toolCallId)) {
          return false;
        }
      }
    }
  }

  return true;
}

/**
 * Find the optimal cut point in `messages` that preserves at least `keepRecentTokens`
 * while strictly adhering to turn-safety and provider role invariants.
 */
export function findCutPoint(
  messages: AgentMessage[],
  keepRecentTokens: number = 20_000,
): CutPointResult {
  if (messages.length <= 4) {
    return { firstKeptIndex: 0, isUserBoundary: true };
  }

  // Find all user turn indices
  const userIndices: number[] = [];
  for (let i = 0; i < messages.length; i++) {
    if (messages[i].role === "user") {
      userIndices.push(i);
    }
  }

  // Walk backwards from newest, accumulating token count
  let accumulatedTokens = 0;
  let targetIndex = 0;

  for (let i = messages.length - 1; i >= 0; i--) {
    accumulatedTokens += estimateMessageTokens([messages[i]]);
    if (accumulatedTokens >= keepRecentTokens) {
      targetIndex = i;
      break;
    }
  }

  // Preference 1: Nearest user turn at or after targetIndex (to ensure we preserve enough tokens,
  // or right before targetIndex if targetIndex is after the last user turn).
  // Look for a user turn that gives a safe cut point.
  let bestCutIndex = -1;

  // Search user turns from newest to oldest
  for (let u = userIndices.length - 1; u >= 0; u--) {
    const idx = userIndices[u];
    // We want a user turn such that we don't discard the whole conversation (idx > 0)
    if (idx > 0 && idx >= targetIndex && isToolCallSafe(messages, idx)) {
      bestCutIndex = idx;
      break;
    }
  }

  // If no user turn at or after targetIndex was found, try the closest user turn before targetIndex
  if (bestCutIndex === -1) {
    for (let u = userIndices.length - 1; u >= 1; u--) {
      const idx = userIndices[u];
      if (isToolCallSafe(messages, idx)) {
        bestCutIndex = idx;
        break;
      }
    }
  }

  if (bestCutIndex > 0) {
    return { firstKeptIndex: bestCutIndex, isUserBoundary: true };
  }

  // Preference 2: If the session has only 1 user turn (e.g. 50 tool turns in one prompt),
  // walk backwards from targetIndex and find the first tool-call safe assistant message.
  for (let i = targetIndex; i < messages.length - 2; i++) {
    if (messages[i].role === "assistant" && isToolCallSafe(messages, i)) {
      return { firstKeptIndex: i, isUserBoundary: false };
    }
  }

  // Fallback: cannot safely compact without breaking turn pairing
  return { firstKeptIndex: 0, isUserBoundary: true };
}

File: apps/server-go/internal/agent/compaction/cutpoint.go
// Cut-point selection: discard old turns without orphaning tool results
// from their calls or breaking provider role alternation. Port of
// apps/server/agent/src/compaction/cut-point.ts.
package compaction

import (
	"github.com/Adelodunpeter25/console/apps/server-go/internal/agent/loop"
)

// CutPointResult is the first-kept index plus boundary quality.
type CutPointResult struct {
	FirstKeptIndex int
	IsUserBoundary bool
}

// IsToolCallSafe reports whether slicing at candidateIndex keeps every
// toolResult paired with its assistant toolCall.
func IsToolCallSafe(history []any, candidateIndex int) bool {
	if candidateIndex <= 0 || candidateIndex >= len(history) {
		return true
	}
	if isToolResultMessage(history[candidateIndex]) {
		return false
	}
	discarded := make(map[string]bool)
	for i := 0; i < candidateIndex; i++ {
		for _, id := range toolCallIDs(history[i]) {
			discarded[id] = true
		}
	}
	for i := candidateIndex; i < len(history); i++ {
		if results, ok := toolResultsOf(history[i]); ok {
			for _, r := range results {
				if discarded[r.ToolCallID] {
					return false
				}
			}
		}
	}
	return true
}

// FindCutPoint locates the first-kept index preserving at least
// keepRecentTokens while honoring protected recent turns and tool pairing.
func FindCutPoint(history []any, keepRecentTokens, minimumRecentTurns int) CutPointResult {
	if keepRecentTokens <= 0 {
		keepRecentTokens = 40000
	}
	if minimumRecentTurns <= 0 {
		minimumRecentTurns = 3
	}
	if len(history) <= 4 {
		return CutPointResult{FirstKeptIndex: 0, IsUserBoundary: true}
	}
	var userIndices []int
	for i, m := range history {
		if isUserMessage(m) {
			userIndices = append(userIndices, i)
		}
	}
	protectedStart := 0
	if len(userIndices) > minimumRecentTurns {
		protectedStart = userIndices[len(userIndices)-minimumRecentTurns]
	}
	accumulated := 0
	targetIndex := 0
	for i := len(history) - 1; i >= 0; i-- {
		accumulated += EstimateMessageTokens(history[i : i+1])
		if accumulated >= keepRecentTokens {
			targetIndex = i
			break
		}
	}
	if protectedStart > 0 && targetIndex > protectedStart {
		targetIndex = protectedStart
	}
	// Preference 1: newest user turn at/after target inside protection.
	for u := len(userIndices) - 1; u >= 0; u-- {
		idx := userIndices[u]
		if idx > 0 && idx >= targetIndex && idx <= protectedStart && IsToolCallSafe(history, idx) {
			return CutPointResult{FirstKeptIndex: idx, IsUserBoundary: true}
		}
	}
	// Preference 2: closest safe user turn before target.
	for u := len(userIndices) - 1; u >= 1; u-- {
		idx := userIndices[u]
		if IsToolCallSafe(history, idx) {
			return CutPointResult{FirstKeptIndex: idx, IsUserBoundary: true}
		}
	}
	// Preference 3: single-user-turn sessions — first safe assistant cut.
	for i := targetIndex; i < len(history)-2; i++ {
		if i > protectedStart {
			continue
		}
		if isAssistantMessage(history[i]) && IsToolCallSafe(history, i) {
			return CutPointResult{FirstKeptIndex: i, IsUserBoundary: false}
		}
	}
	return CutPointResult{FirstKeptIndex: 0, IsUserBoundary: true}
}

func isToolResultMessage(m any) bool {
	switch msg := m.(type) {
	case loop.ToolResultMessage:
		return msg.Role == loop.RoleToolResult
	case *loop.ToolResultMessage:
		return msg != nil && msg.Role == loop.RoleToolResult
	}
	return false
}

func isAssistantMessage(m any) bool {
	switch msg := m.(type) {
	case loop.AssistantMessage:
		return msg.Role == loop.RoleAssistant
	case *loop.AssistantMessage:
		return msg != nil && msg.Role == loop.RoleAssistant
	}
	return false
}

func toolCallIDs(m any) []string {
	var ids []string
	var content []any
	switch msg := m.(type) {
	case loop.AssistantMessage:
		content = msg.Content
	case *loop.AssistantMessage:
		if msg == nil {
			return nil
		}
		content = msg.Content
	default:
		return nil
	}
	for _, part := range content {
		if call, ok := part.(loop.ToolCallPart); ok {
			ids = append(ids, call.Call.ID)
		}
	}
	return ids
}
